### PR TITLE
Add: basic encryption, message models, Forget, Program

### PR DESCRIPTION
**Pull request by Rgascoin**
###  What's new:

**Major changes:**
- Message:
  - Change message model definition according to [aleph-message](https://github.com/aleph-im/aleph-message)
  - Add Forget message
  - Add Program message
  - Update Store message (store get a Buffer or a Blob now)

- Account:
  -  Add Encryption secured message encryption for curve secp256k1
  -  Add a beta encryption for curve ed25519

**Minor changes:**
- Readme:
    - Updated of features
- Test:
    - Add tests for new features and fix
- Message: 
  - change the way getAggregate() can be called to simplify its usage

### CODE REVIEW

File: src/accounts/ethereum.ts [hunk 1]
```diff
@@ -14,8 +16,55 @@ export class ETHAccount extends Account {
...
+        let result: Buffer | string;
+
+        if (as_hex && typeof encryptedContent === "string") encryptedContent = Buffer.from(encryptedContent, "hex");
+        else if (typeof encryptedContent === "string") encryptedContent = Buffer.from(encryptedContent);
+
+        const secret = this.wallet.privateKey;
+        result = secp256k1_decrypt(secret, encryptedContent);
+
+        if (result === null) throw new Error("could not decrypt");
+        if (as_string) result = result.toString();
```
Comment (PtitLuca) [inline src/accounts/ethereum.ts:66]: Are you sure about this ? Does the content always be a string ?

File: src/messages/create/publish.ts [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import shajs from "sha.js";
 
-import { BaseMessage, StorageEngine } from "../message";
+import { BaseMessage, ItemType } from "../message";
 import axios from "axios";
+import FormDataNode from "form-data";
```
Comment (PtitLuca) [inline src/messages/create/publish.ts:5]: Please make sure this works in the browser.
You can test this using cypress.

File: src/messages/create/publish.ts [hunk 3]
```diff
@@ -75,20 +76,28 @@ async function PushToStorageEngine<T>(configuration: PushConfiguration<T>): Prom
...
+    const isBrowser = typeof FormData !== "undefined";
+    let form: FormDataNode | FormData;
 
-    form.append("file", configuration.file);
+    if (isBrowser) {
+        form = new FormData();
+        form.append("file", new Blob([configuration.file]));
+    } else {
+        form = new FormDataNode();
+        form.append("file", configuration.file, "usageName.txt");
```
Comment (PtitLuca) [inline src/messages/create/publish.ts:91]: Is it a normal behaviour to have this hardcoded string ?

File: src/messages/message.ts [hunk 4]
```diff
@@ -1,39 +1,116 @@
...
+    aggregate = "AGGREGATE",
+    store = "STORE",
+    program = "PROGRAM",
+    forget = "FORGET",
+}
+
+export enum ItemType {
+    inline = "inline",
+    storage = "storage",
+    ipfs = "ipfs",
```
Comment (PtitLuca) [inline src/messages/message.ts:43]: Are you merging a storage engine and a storage method ?

File: src/messages/create/publish.ts (hunk 3 above)
Comment (Rgascoin) [reply src/messages/create/publish.ts]: This last parameter represents the fileName according to the [documentation](https://developer.mozilla.org/fr/docs/Web/API/FormData/append).

It does not appear to be mandatory, but I got an error 500 when I don't mention it.
So I changed `"usageName.txt"` by the default value `"File"`=> 709c021

File: src/messages/message.ts (hunk 4 above)
Comment (Rgascoin) [reply src/messages/message.ts:43]: As far as I know nop, I updated this structure to fit with [aleph-message](https://github.com/aleph-im/aleph-message/blob/main/aleph_message/models/__init__.py)

File: src/accounts/nuls2.ts [hunk 5]
```diff
@@ -22,8 +24,27 @@ export class NULS2Account extends Account {
...
+    }
+
+    /**
+     * Decrypt a given content using a NULS2 account.
+     *
+     * @param encryptedContent The encrypted content to decrypt.
+     */
+    decrypt(encryptedContent: Buffer): Buffer {
+        const secret = this.privateKey;
+        return secp256k1_decrypt(secret, encryptedContent);
```
Comment (PtitLuca) [inline src/accounts/nuls2.ts:47]: No need to create a temp. variable here

File: src/messages/program/publish.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        },
+        resources: {
+            vcpus: 1,
+            memory: configuration.memory ? configuration.memory : 128,
+            seconds: 30,
+        },
+        runtime: {
+            ref: configuration.runtime
+                ? configuration.runtime
+                : "bd79839bf96e595a06da5ac0b6ba51dea6f7e2591bb913deccded04d831d29f4",
```
Comment (PtitLuca) [inline src/messages/program/publish.ts:68]: What does this ref. is refering to ?

**Review by PtitLuca — CHANGES_REQUESTED** [review]
Please take a look at my comments and feel free to merge after changes :rocket:

File: src/accounts/nuls2.ts (hunk 5 above)
Comment (Rgascoin) [reply src/accounts/nuls2.ts]: Fixed in [c099ae5](https://github.com/aleph-im/aleph-sdk-ts/pull/10/commits/c099ae559076917cfc9b9007f866d9dbe09bce07)

File: src/messages/program/publish.ts (hunk 6 above)
Comment (Rgascoin) [reply src/messages/program/publish.ts:68]: runtime refers to the `root filesystem` used by the Aleph VM
and the ref param is a sort of path to them, you can find the official Aleph runtimes [here](https://github.com/aleph-im/aleph-vm/tree/main/runtimes)

**Review by PtitLuca — APPROVED** [review]
LGTM :rocket:

**Comment by Rgascoin** [comment]
@geraldinefiser 😉

**Comment by Rgascoin** [comment]
@geraldinefiser 😉

**Comment by geraldinefiser** [comment]
@Rgascoin Updates to the docs made here: https://aleph-im.gitbook.io/ts-sdk/v/v-2.2.2/

**Pull request merged**